### ISSUE TITLE
Kernel dies during post-training wrap-up when training more than 500 iterations

### ISSUE THREAD

**Issue by JivanRoquet** [dmlc/xgboost#7913]
Using XGBoost 1.6.0 with a multi-class classification problem involving 6000+ target classes, about 500 input features and about 2M training rows.

Training on a p3.16xlarge 8-GPU machine with Dask on 16 partitions works:

- fine for up to about 500 iterations
- causes the kernel to die if training for more than about 500 iterations (e.g. 600, 700, 800...)

In both cases the training goes fine up to the end, but in the second case, the kernel dies after the last iteration, during what seems to be some internal post-processing done within the `fit()` method.

In both cases, all 8 GPUs seem to be working properly with their respective memory filled up to about 75% during the whole training.

Relevant portion of the code:

```python
model = xgb.dask.DaskXGBClassifier(
    max_depth=16,
    n_estimators=500,
    learning_rate=0.07,
    max_bin=128,
    reg_alpha=0.1,
    colsample_bytree=0.5,
    colsample_bylevel=0.8,
    colsample_bynode=0.9,
    subsample=0.2,
    verbosity=1,
    silent=None,
    grow_policy="lossguide",
    objective="multi:softproba",
    tree_method='gpu_hist',
    n_jobs=-1,
    eval_metric=my_eval_metric,
    booster="gbtree",
    callbacks=[my_callback],
    predictor="cpu_predictor",
)

X_train_da = dask.dataframe.from_pandas(pd.DataFrame(X_train), 16)
y_train_da = dask.dataframe.from_pandas(pd.Series(y_train), 16)
X_eval_da = dask.dataframe.from_pandas(pd.DataFrame(X_eval), 16)
y_eval_da = dask.dataframe.from_pandas(pd.Series(y_eval), 16)
weights_da = dask.dataframe.from_pandas(pd.Series(weights), 16)

with LocalCUDACluster() as cluster:
    with Client(cluster) as client:
       model.fit(    #   <---- Kernel dies during this call, after all training steps have successfully ran
            X_train_da,
            y_train_da,
            eval_set=[(X_eval_da, y_eval_da)],
            sample_weight=weights_da,
        )
```

Again, this code works perfectly fine up to about 500 training iterations (n_estimators=500).

With more than about 500 iterations, the kernel dying happens at the end of the call to `model.fit()` during the final wrap-up, after all training steps successfully terminated and their logging information (train/val loss...) having been successfully printed, including the very last one.

**Comment by trivialfis** [comment]
Is it possible that there's OOM error?

**Comment by JivanRoquet** [comment]
@trivialfis thanks for pointing this. I initially didn't think it could be RAM-related, as this is a 480Gb RAM instance with about 150Gb used during training.

It's possible that there's a sudden RAM usage spike at the very end during the wrap-up, I'll need to check what happens precisely at this moment, see if there's a significant spike.

**Comment by JivanRoquet** [comment]
@trivialfis it was definitely and absolutely an OOM error.

This is the profile of RAM usage (in GB) right after the training ends (first red arrow) up until the `fit()` method returns (second red arrow).

<img width="774" alt="Screenshot 2022-05-18 at 16 29 53" src="https://user-images.githubusercontent.com/6452675/169082322-eaaf4749-03eb-4fb2-9867-b1d3ad1fe6bb.png">

Most of the training time happens using about 150GB of RAM, fairly constantly, and then there's a spike at 360GB. The delay between the two red arrows is approximately 12 minutes.

Thanks for initially pointing to this eventuality, that was very helpful.

**Comment by JivanRoquet** [comment]
Size of training data is roughly 2M rows with 550 columns, mostly floats. I've actually just realized that the `dtype` of the training array is `float64`. Maybe changing it to `float32` would ease the problem.

**Comment by trivialfis** [comment]
I'm not sure what's taking up the memory since you are using gpu hist.

> X_train_da = dask.dataframe.from_pandas(pd.DataFrame(X_train), 16)

This is not a good practice and might cost significant amount of memory, see https://xgboost.readthedocs.io/en/stable/tutorials/dask.html#memory-usage .

**Comment by trivialfis** [comment]
@JivanRoquet You can try to replace the `from_pandas` by using dask.dataframe to load the data directly.

**Comment by JivanRoquet** [comment]
@trivialfis thanks for the tip. Could doing this have a link with us currently observing no speed improvement when the training is done on 1 GPU (without Dask) vs 4 or 8 GPUs (with Dask)?

Essentially currently we need e.g. 8 GPU in parallel only because when using 4 GPU, the GPUs run out of memory at the beginning of the training. But we don't observe any speed improvement by doing so, just that the GPU OOM doesn't happen.

**Comment by trivialfis** [comment]
> Could doing this have a link with us currently observing no speed improvement when the training is done on 1 GPU (without Dask) vs 4 GPUs vs 8 GPUs (with Dask)?

I can't be sure. It's highly dependent on the data and the cluster you are using. We are currently trying to reduce the communication overhead in XGBoost, which might help.

**Comment by JivanRoquet** [comment]
For what it's worth, I've changed the Dask dataframe loading approach by directly reading from Parquet files and the behaviour described above still holds (big RAM usage increase on wrapping up).

**Comment by trivialfis** [comment]
Yup, that's a separate issue that still needs some investigation. But I hope that by changing the data loading approach you can avoid an OOM error.

**Comment by trivialfis** [comment]
I haven't tested it yet, but an initial guess can show that the memory surge can be expected. Take a conservative estimation that each tree node takes up 64 bytes,  a 16 depth dense tree has about 60k nodes (we don't usually grow a fully dense tree, just to give an upper bound), 6000x500 trees with models gathered from 8 workers, that's 1.4TB!
